pkg/controller/sparkapplication/submission_cache.go

```diff
@@ -0,0 +1,66 @@
+package sparkapplication
+
+/*
+A simple cache implementation that stores keys with a time-to-live (TTL) value.
+*/
+
+import (
+	"sync"
+	"time"
+)
+
+type item struct {
+	expiration int64
+}
+
+type SubmissionCache struct {
+	ttl   time.Duration
+	items map[string]*item
+	mu    sync.RWMutex
+}
+
+func NewSubmissionCache(ttl time.Duration) *SubmissionCache {
+	c := &SubmissionCache{
+		ttl:   ttl,
+		items: make(map[string]*item),
+	}
+
+	go c.startEviction()
+
+	return c
+}
+
+func (c *SubmissionCache) Set(key string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.items[key] = &item{
+		expiration: time.Now().Add(c.ttl).UnixNano(),
+	}
+}
+
+func (c *SubmissionCache) Exist(key string) bool {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	item, found := c.items[key]
+	if !found || item.expiration < time.Now().UnixNano() {
+		return false
+	}
+
+	return true
+}
+
+func (c *SubmissionCache) startEviction() {
+	for {
+		time.Sleep(c.ttl)
+
+		c.mu.Lock()
+		for key, item := range c.items {
+			if time.Now().UnixNano() > item.expiration {
+				delete(c.items, key)
+			}
+		}
+		c.mu.Unlock()
```

Comment:
Why not defer this one?

Reply:
nevermind, I read the code too quickly. you'd have to put the inner loop in a sub function to be able to defer.

Reply:
This is a good point it's always better to run:
```
c.mu.Lock()
defer c.mu.Unlock()
```
in that use case to prevent the case of deadlock if something goes wrong (even if the code seems perfectly safe)

Reply:
it's inside a loop, that's why I don't use defer here